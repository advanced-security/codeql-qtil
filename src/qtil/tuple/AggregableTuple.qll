private import qtil.parameterization.SignatureTypes
private import qtil.parameterization.SignaturePredicates
private import qtil.tuple.StringTuple as CustomStringTuple
private import qtil.strings.Chars
private import qtil.inheritance.Instance
private import codeql.util.Boolean

class StringTuple = CustomStringTuple::StringTuple<Chars::comma/0>::Tuple;

/**
 * A module that allows multiple values to be aggregated at the same time, where each value
 * (including the aggregated value) acts like a tuple.
 *
 * The tuple may contain any number of the following types of columns:
 * - `string` columns, which are concatenated with a separator
 * - `int` columns, which are summed
 *
 * Additionally, the unique values of each column can be counted, and the total number of unique
 * aggregated tuples can be counted.
 *
 * This can be useful for writing generic code where a module may wish to perform an unknown number
 * of aggregations in a context where it cannot perform the aggregation for itself.
 *
 * Each value to be aggregated should be of type `AggregableTuple::Piece`, and pieces should be
 * aggregated with `concat(Piece p | p, ",")`, as the underlying representation is a comma
 * -separated string (a `StringTuple`).
 *
 * After aggregation, the result should be cast to a `AggregableTuple::Sum` to access the
 * aggregated values of each column.
 *
 * Note: This will not be as performant as individual aggregations, and should only be used in cases
 * where a single aggregation is not practical.
 *
 * Example usage:
 * ```ql
 * // What values a "person" may aggregate over defined here:
 * AggregableTuple::Piece personAggregant(Person p) {
 *   result = AggregableTuple::initString(p.name)
 *            .appendInt(p.age)
 * }
 *
 * // A usage of that aggregation can be defined separately:
 * predicate useAggregation(AggregableTuple::Sum<two/0>::Sum aggregated) {
 *   exists(int counted, string names, int totalAge |
 *     counted = aggregated.getCountTotal() and
 *     names = aggregated.getAsJoinedString(0, ",") and
 *     totalAge = aggregated.getAsSummedInt(1) and
 *     // Use `counted`, `names`, and `totalAge` as needed
 *   )
 * }
 * ```
 */
module AggregableTuple {
  /**
   * Begin the construction of a new piece of an aggregable tuple with a `string` column.
   *
   * Sets the first column of this tuple to be the given `string` value. The `Piece`
   * returned by this predicate can have additional columns appended to it of any type.
   */
  bindingset[s]
  Piece initString(string s) { result = s }

  /**
   * Begin the construction of a new piece of an aggregable tuple with an `int` column.
   *
   * Sets the first column of this tuple to be the given `int` value. The `Piece`
   * returned by this predicate can have additional columns appended to it of any type.
   */
  bindingset[i]
  Piece initInt(int i) { result = i.toString() }

  /**
   * A piece of an aggregable tuple, which can be used to aggregate multiple values at the same
   * time.
   *
   * This class can be built up one column at a time, beginning with one of the predicates `asInc`,
   * `asString`, or `asInt`. Additional columns can be appended to the piece using the `appendInc`,
   * `appendString`, or `appendInt` predicates.
   *
   * After all of the columns have been appended, the piece can be aggregated with
   * `concat(Piece p | p, ",")`. Then the result can be cast to `AggregableTuple::Sum` to access the
   * aggregated values of each column.
   */
  bindingset[this]
  class Piece extends InfInstance<StringTuple>::Type {
    bindingset[this, s]
    Piece appendString(string s) { result = inst().append(s) }

    bindingset[this, i]
    Piece appendInt(int i) { result = inst().append(i.toString()) }
  }

  module Sum<Nullary::Ret<int>::pred/0 columns> {
    bindingset[this]
    class Sum extends InfInstance<StringTuple>::Type {
      bindingset[this]
      int getCountTotal() { result = inst().size() / columns() }

      /**
       * Since the underlying representation is a comma-separated string, the ith value of
       * the nth column can be found at the index `i * columns() + n`.
       *
       * This predicate returns all such indexes for the nth column.
       */
      bindingset[this]
      int getARawColumnValueIndex(int colIdx) {
        colIdx in [0 .. columns()] and
        exists(int rowIdx |
          rowIdx = [0 .. getCountTotal() - 1] and
          result = rowIdx * columns() + colIdx
        )
      }

      /**
       * Get all of the raw string values for the nth column of aggregated tuples.
       */
      bindingset[this]
      string getARawColumn(int colIdx) {
        colIdx in [0 .. columns()] and
        result = inst().get(getARawColumnValueIndex(colIdx))
      }

      bindingset[this]
      int countColumn(int colIdx) {
        colIdx in [0 .. columns()] and
        result = count(string item | item = getARawColumn(colIdx))
      }

      /**
       * Get the nth column of aggregated tuples, treated as strings and joined with the given
       * separator.
       */
      bindingset[this, sep]
      string getAsJoinedString(int colIdx, string sep) {
        colIdx in [0 .. columns()] and
        result = concat(string item | item = getARawColumn(colIdx) | item, sep)
      }

      /**
       * Get the nth column of aggregated tuples, treated as integers and summed.
       */
      bindingset[this]
      int getAsSummedInt(int colIdx) {
        colIdx in [0 .. columns()] and
        result = sum(int item | item = getARawColumn(colIdx).toInt())
      }
    }
  }
}
