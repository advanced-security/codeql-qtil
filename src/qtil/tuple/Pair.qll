private import qtil.parameterization.SignatureTypes
private import qtil.parameterization.SignaturePredicates

/**
 * A module that provides a pair of values of type A and B.
 *
 * In `qtil` parlance, a "Pair" is a specific set of two types, A and B, constrained by some
 * predicate, a "Tuple" has three or more types/values, and a "Product" is the combination of
 * types/values which are not constrained by a predicate.
 *
 * To use this module, write a predicate that holds for the two types you want to constrain, and
 * instantiate this module like so:
 *
 * ```ql
 * import qtil.tuple.Pair
 *
 * predicate myPredicate(int a, string b) {
 *   a = [0..10] and b = a.toString()
 * }
 *
 * class MyIntStringPair = Pair<int, string, myPredicate>::Pair;
 * ```
 *
 * Each tuple type exposes the `getFirst()` and `getSecond()` member predicates, which retrieves
 * the first and second values of the pair, respectively.
 */
module Pair<InfiniteStringableType A, InfiniteStringableType B, Binary<A, B>::pred/2 constraint> {
  private newtype TAll = TSome(A a, B b) { constraint(a, b) }

  class Pair extends TAll {
    A getFirst() { this = TSome(result, _) }

    B getSecond() { this = TSome(_, result) }

    string toString() { result = getFirst().toString() + ", " + getSecond().toString() }
  }
}
