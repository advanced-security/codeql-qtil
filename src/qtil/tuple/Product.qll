private import qtil.parameterization.SignatureTypes

/**
 * A module that provides the product of values of type A and B.
 *
 * *Caution*: Product types can have a large number of values and are not suitable for many use
 * cases. Carefully consider if using a `Pair` or `Tuple` type is more appropriate before using
 * this module.
 *
 * In `qtil` parlance, a "Pair" is a specific set of two types, A and B, constrained by some
 * predicate, a "Tuple" has three or more types/values, and a "Product" is the combination of
 * types/values which are not constrained by a predicate.
 *
 * To use this module, instantiate the module with a set of finite types. Infinite types are not
 * supportable.
 *
 * ```ql
 * import qtil.tuple.Product
 *
 * class ABProduct = Product2<A, B>::Product;
 * ```
 *
 * Each product exposes the `getFirst()` and `getSecond()` member predicates, which retrieves
 * the first and second values of the product, respectively.
 */
module Product2<FiniteStringableType A, FiniteStringableType B> {
  private newtype TAll = TSome(A a, B b)

  class Product extends TAll {
    A getFirst() { this = TSome(result, _) }

    B getSecond() { this = TSome(_, result) }

    string toString() { result = getFirst().toString() + ", " + getSecond().toString() }
  }
}
