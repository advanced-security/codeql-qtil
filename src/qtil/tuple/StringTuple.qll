private import qtil.strings.Char
private import qtil.strings.Escape
private import qtil.parameterization.Finalize
private import qtil.parameterization.SignaturePredicates
private import qtil.parameterization.SignatureTypes
private import qtil.inheritance.UnderlyingString
private import qtil.inheritance.Instance
private import qtil.list.ListBuilder
private import qtil.tuple.Pair
private import qtil.tuple.Tuple

/**
 * A tuple that contains only string values, such as a csv row.
 *
 * This module can be useful to create infinite types with some basic structure.
 */
module StringTuple<Nullary::Ret<Char>::pred/0 separator> {
  private import SeparatedEscape<separator/0, emptyEscapeMap/2> as Escape

  bindingset[this]
  class Tuple extends Final<UnderlyingString>::Type {
    bindingset[this, idx]
    string get(int idx) { result = Escape::split(str(), idx) }

    bindingset[this, item]
    Tuple append(string item) { result = str() + separator() + Escape::escapeItem(item) }

    bindingset[this]
    int size() { result = count(int idx | exists(Escape::split(str(), idx))) }
  }

  // Import constructors `Tuple of2(...)` to `Tuple of8(...)` from SeparatedList module.
  // Currently adds ~4sec on its own to E2E, slower in problems() alone:
  import TypedListBuilderOf<separator/0, Tuple, string, Escape::escapeItem/1>
  // Include the `Typed` submodule, usable as `StringTuple<sep>::Typed<T, fromString>::List`.
  import TypedStringTuple<separator/0>
}

private module TypedStringTuple<Nullary::Ret<Char>::pred/0 separator> {
  module Typed<InfiniteType T, Unary<string>::Ret<T>::bindInputOutput/1 fromString> {
    bindingset[item]
    bindingset[result]
    private string toString(T item) { item = fromString(result) }

    bindingset[this]
    class Tuple extends InfInstance<StringTuple<separator/0>::Tuple>::Type {
      bindingset[this, item]
      Tuple append(T item) { result = inst().append(toString(item)) }

      bindingset[this]
      int size() { result = inst().size() }

      bindingset[this, idx]
      T get(int idx) { result = fromString(inst().get(idx)) }
    }

    // Import constructors `Tuple of2(...)` to `Tuple of8(...)` from ListBuilder module.
    import TypedListBuilderOf<separator/0, Tuple, T, toString/1>
  }
}
