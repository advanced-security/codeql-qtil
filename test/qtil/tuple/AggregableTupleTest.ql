import qtil.tuple.AggregableTuple
import qtil.testing.Qnit

class TestInitString extends Test, Case {
  override predicate run(Qnit test) {
    if AggregableTuple::initString("test") = "test"
    then test.pass("Correctly initialized a string")
    else test.fail("initString did not initialize correctly")
  }
}

class TestInitInt extends Test, Case {
  override predicate run(Qnit test) {
    if AggregableTuple::initInt(42) = "42"
    then test.pass("Correctly initialized an integer")
    else test.fail("initInt did not initialize correctly")
  }
}

class TestInitAppendString extends Test, Case {
  override predicate run(Qnit test) {
    if
      AggregableTuple::initString("test1").appendString("test2") = "test1,test2" and
      AggregableTuple::initString("test1").appendString("test2").appendString("test3") =
        "test1,test2,test3"
    then test.pass("Correctly appended multiple strings")
    else test.fail("appendString did not append multiple strings correctly")
  }
}

class TestInitAppendInt extends Test, Case {
  override predicate run(Qnit test) {
    if
      AggregableTuple::initInt(1).appendInt(2) = "1,2" and
      AggregableTuple::initInt(1).appendInt(2).appendInt(3) = "1,2,3"
    then test.pass("Correctly appended multiple integers")
    else test.fail("appendInt did not append multiple integers correctly")
  }
}

class TestInitAppendMixed extends Test, Case {
  override predicate run(Qnit test) {
    if
      AggregableTuple::initString("test").appendInt(42) = "test,42" and
      AggregableTuple::initInt(42).appendString("test") = "42,test"
    then test.pass("Correctly appended mixed types")
    else test.fail("appendMixed did not append mixed types correctly")
  }
}

int one() { result = 1 }

class TestConcatSingleString extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initString("test") | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsJoinedString(0, ",") = "test"
    then test.pass("Correctly concatenated single string")
    else test.fail("concat did not concatenate single string correctly")
  }
}

class TestSumSingleInt extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initInt(42) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsSummedInt(0) = 42
    then test.pass("Correctly summed single integer")
    else test.fail("concat did not sum single integer correctly")
  }
}

class TestJoinMultipleStrings extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initString(["test1", "test2"]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsJoinedString(0, ",") = "test1,test2" and
      concat(string s | s = AggregableTuple::initString(["test1", "test2", "test3"]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsJoinedString(0, ",") = "test1,test2,test3"
    then test.pass("Correctly joined multiple strings")
    else test.fail("concat did not join multiple strings correctly")
  }
}

class TestSumMultipleIntegers extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initInt([1, 2]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsSummedInt(0) = 3 and
      concat(string s | s = AggregableTuple::initInt([1, 2, 3]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .getAsSummedInt(0) = 6
    then test.pass("Correctly summed multiple integers")
    else test.fail("concat did not sum multiple integers correctly")
  }
}

class TestCountMultipleStrings extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initString(["test1", "test2"]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .countColumn(0) = 2 and
      concat(string s | s = AggregableTuple::initString(["test1", "test2", "test3"]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .countColumn(0) = 3
    then test.pass("Correctly counted multiple strings")
    else test.fail("concat did not count multiple strings correctly")
  }
}

class TestCountMultipleIntegers extends Test, Case {
  override predicate run(Qnit test) {
    if
      concat(string s | s = AggregableTuple::initInt([1, 2]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .countColumn(0) = 2 and
      concat(string s | s = AggregableTuple::initInt([1, 2, 3]) | s, ",")
          .(AggregableTuple::Sum<one/0>::Sum)
          .countColumn(0) = 3
    then test.pass("Correctly counted multiple integers")
    else test.fail("concat did not count multiple integers correctly")
  }
}

int four() { result = 4 }

class TestAggregateMultiColumnPieces extends Test, Case {
  override predicate run(Qnit test) {
    if
      exists(AggregableTuple::Sum<four/0>::Sum summed |
        summed =
          concat(string s |
            s =
              AggregableTuple::initString(["test1", "test2"])
                  .appendInt([1, 2, 3])
                  .appendInt([2, 3, 4])
                  .appendString(["test3", "test4"])
          |
            s, ","
          ) and
        summed.getAsJoinedString(0, ",") = "test1,test2" and
        summed.getAsSummedInt(1) = 6 and
        summed.getAsSummedInt(2) = 9 and
        summed.getAsJoinedString(3, ",") = "test3,test4"
      )
    then test.pass("Correctly aggregated multi-column pieces")
    else test.fail("concat did not aggregate multi-column pieces correctly")
  }
}
