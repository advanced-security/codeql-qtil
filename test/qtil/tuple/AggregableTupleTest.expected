| All 12 tests passed. |
